src/executables/master_heartbeat.c

```diff
@@ -6389,6 +6687,48 @@ hb_check_ping (const char *host)
   return HB_PING_SUCCESS;
 }
 
+/*
+ * hb_check_tcp_ping -
+ *   return : int
+ *
+ */
+static int
+hb_check_tcp_ping (const char *host, int port, int timeout)
+{
+  char buf[128];
+  HB_NODE_ENTRY *node;
+  SOCKET sfd = INVALID_SOCKET;
+
+  /* If host is in the cluster node, then skip to check */
+  for (node = hb_Cluster->nodes; node; node = node->next)
+    {
+      if (are_hostnames_equal (host, node->host_name))
+	{
+	  /* PING Host is same as cluster's host name */
+	  snprintf (buf, sizeof (buf), "Useless TCP PING host name %s", host);
+	  MASTER_ER_SET (ER_ERROR_SEVERITY, ARG_FILE_LINE, ER_HB_NODE_EVENT, 1, buf);
+	  return HB_PING_USELESS_HOST;
+	}
+    }
+
+  sfd = css_tcp_client_open_with_timeout (host, port, timeout);
```

Comment:
Doesn't it need a TCP host that listens and accepts on the port?

Reply:
The tcp ping(ping) check acts as a client. There is a server that allows tcp connection to the outside(I tested 22 ssh port), and cubrid operates as a client. Also, check the network status by connecting/disconnecting rather than checking by message communication.

Reply:
It means to just connect to any kind of TCP server listening on the port. I didn't think of this way. Thank you. 
However, this way depends on a random external application that listens on the port. In other words, the heartbeat is not reliable itself. It's different from ICMP in that the ICMP is designed for the purpose. Is it acceptable?

Reply:
If I'm correct, it seems pretty vulnerable so It should be used carefully. This should be described as a warning in the manual.

Reply:
@wrlawodms The TCP ip/port is selected by the dba. If the unreliable ip/port is selected, then the system believes the unreliable result of a TCP ping even if there are no network issues. I will leave a commet(NOTE) in the manual.